File: 156_Merge_Intervals.cpp
/*
156. Merge Intervals
Given a collection of intervals, merge all overlapping intervals.

Example
Given intervals => merged intervals:

[                     [
  (1, 3),               (1, 6),
  (2, 6),      =>       (8, 10),
  (8, 10),              (15, 18)
  (15, 18)            ]
]

Challenge
O(n log n) time and O(1) extra space.


Author: Yuan Li
Date: 10/29/2018 
Difficulty: Easy
*/

/**
 * Definition of Interval:
 * classs Interval {
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param intervals: interval list.
     * @return: A new interval list.
     */
    vector<Interval> merge(vector<Interval> &intervals) {
        // write your code here
        sort(intervals.begin(), intervals.end(), 
            [](const Interval &lhs, const Interval &rhs) {
                return lhs.start < rhs.start;    
            });
            
        vector<Interval> res;
        for (auto intv : intervals) {
            if (res.empty() || res.back().end < intv.start) {
                res.push_back(intv);
            } else {
                res.back().end = max(res.back().end, intv.end);
            }
        }
        
        return res;
    }
};

File: 156_Merge_Intervals.py
#!/usr/bin/python3

'''
156. Merge Intervals
Given a collection of intervals, merge all overlapping intervals.

Example
Given intervals => merged intervals:

[                     [
  (1, 3),               (1, 6),
  (2, 6),      =>       (8, 10),
  (8, 10),              (15, 18)
  (15, 18)            ]
]

Challenge
O(n log n) time and O(1) extra space.


Author: Yuan Li
Date: 10/29/2018 
Difficulty: Easy
'''

"""
Definition of Interval.
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Solution:
    """
    @param intervals: interval list.
    @return: A new interval list.
    """
    def merge(self, intervals):
        # write your code here
        intervals.sort(key=lambda x: x.start)
        res = []
        for intv in intervals:
            if not res or res[-1].end < intv.start:
                res.append(intv)
            else:
                res[-1].end = max(res[-1].end, intv.end)
        
        return res

